### PR TITLE
Improve DateRange negative case handling

### PR DESCRIPTION
**Pull request by mgrassotti**
Trello: https://trello.com/c/YVhH1JFx/780-3-improve-daterange-corner-case-handling

If the end date is initialised to a date before the begins date,
`DateRange` silently sets it to the day before the begins date. This was
probably done to make it return 0 number of days. This is a strange
behaviour and could lead to unexpected results.

Instead of automatically set the end date, this PR allows the end date to
be before the begins date and makes DateRange return the exact amount of
negative number of days. The `empty?` method is also set to return
true in a case of negative or zero number of days.

This avoids to break the other related calculations (e.g.
`gap_between`) and doesn’t make changes to the end date initially set.

### CODE REVIEW

File: lib/smart_answer/date_range.rb [hunk 1]
```diff
@@ -72,7 +71,7 @@ def infinite?
     end
 
     def empty?
```
Comment (andrewgarner) [inline lib/smart_answer/date_range.rb]: I'm not sure I agree with this logic. Days fall between the two dates so it is wrong to say the range is empty.

Comment (mgrassotti) [reply lib/smart_answer/date_range.rb]: @andrewgarner sorry I don't understand, which days fall between the 2 dates?

Comment (andrewgarner) [reply lib/smart_answer/date_range.rb]: I was thinking in regards to ends on being before begins on. However, if were to use [Ruby's Range](https://ruby-doc.org/core-2.2.0/Range.html) library then it would be correct to say to is empty.

```
>> require 'date'
true
>> today = Date.today
#<Date: 2017-11-27 ((2458085j,0s,0n),+0s,2299161j)>
>> october = Date.new(2017, 10, 1)
#<Date: 2017-10-01 ((2458028j,0s,0n),+0s,2299161j)>
>> (october..today).to_a
[
    [ 0] #<Date: 2017-10-01 ((2458028j,0s,0n),+0s,2299161j)>,
    [ 1] #<Date: 2017-10-02 ((2458029j,0s,0n),+0s,2299161j)>,
    [ 2] #<Date: 2017-10-03 ((2458030j,0s,0n),+0s,2299161j)>,
    [ 3] #<Date: 2017-10-04 ((2458031j,0s,0n),+0s,2299161j)>,
    # ...
    [57] #<Date: 2017-11-27 ((2458085j,0s,0n),+0s,2299161j)>
]
>> (today..october).to_a
[]
```

Perhaps we should delegate `empty?` to a range?

Comment (mgrassotti) [reply lib/smart_answer/date_range.rb]: @andrewgarner but, if I understood well, when you define the range as `(october..today)` you are defining a different range then `(today..october)`. The first is not empty while the lattest is empty. So it should be coherent also with our implementation, isn't it?

**Comment by chao-xian** [comment]
@mgrassotti from conversation earlier with @andrewgarner to help clarify:


> Thinking was that in Ruby there is a `Range` type
> so `[1...10]` is a range of numbers.
> so if we're going to have a `DateRange` it should act like a specialist type of `Range`.
> so if you instantiate a new `DateRange`, internally it would hold those dates as a `Range` and you could delegate `empty?` to it.
> It feels like we should have a class that is more specific if we have rules about start date succeeding the end date etc

**Comment by mgrassotti** [comment]
@chao-xian what about thinking about this in another PR? It feels to me like a refactoring that falls outside this PR objectives.

**Comment by chao-xian** [comment]
@mgrassotti I think we should have a test that covers the situation described in the Trello card: 
> If you try to set an end date before the start date, it sets your end date to the day before your start date, which is _still_ before your start date, but not the date you tried to set.

**Comment by pmanrubia** [comment]
Would it be possible to add a few test to exercise all the edge cases and make them pass (first commit), then in upcoming commits refactor it so it is easier to read?

**Comment by mgrassotti** [comment]
@chao-xian done 👍

**Comment by mgrassotti** [comment]
@pmanrubia I tried to do that, let me know if that could be ok. Thanks

File: lib/smart_answer/date_range.rb [hunk 2]
```diff
@@ -72,7 +71,7 @@ def infinite?
...
-      number_of_days == 0
+      number_of_days <= 0
```
Comment (alecgibson) [inline lib/smart_answer/date_range.rb]: How about `number_of_days.positive?` ?

Comment (mgrassotti) [reply lib/smart_answer/date_range.rb]: 👍

Comment (mgrassotti) [reply lib/smart_answer/date_range.rb]: @alecgibson or did you mean `!number_of_days.positive?`?
